packages/react-native/ReactCommon/react/nativemodule/webperformance/NativePerformanceObserver.cpp

```diff
@@ -24,75 +26,101 @@ NativePerformanceObserverModuleProvider(
-    jsi::Runtime& /*rt*/,
-    PerformanceEntryType entryType) {
-  auto reporter = PerformanceEntryReporter::getInstance();
+double NativePerformanceObserver::getDroppedEntriesCount(
+    jsi::Runtime& rt,
+    jsi::Object observerObj) {
+  auto observer = std::dynamic_pointer_cast<PerformanceObserver>(
+      observerObj.getNativeState(rt));
+
+  if (!observer) {
+    return 0;
+  }
 
-  reporter->stopReporting(entryType);
+  return observer->getDroppedEntriesCount();
 }
 
-void NativePerformanceObserver::setIsBuffered(
-    jsi::Runtime& /*rt*/,
-    const std::vector<PerformanceEntryType> entryTypes,
-    bool isBuffered) {
-  for (const PerformanceEntryType entryType : entryTypes) {
-    PerformanceEntryReporter::getInstance()->setAlwaysLogged(
-        entryType, isBuffered);
+void NativePerformanceObserver::observe(
+    jsi::Runtime& rt,
+    jsi::Object observerObj,
+    NativePerformanceObserverObserveOptions options) {
+  auto observer = std::dynamic_pointer_cast<PerformanceObserver>(
+      observerObj.getNativeState(rt));
+
+  if (!observer) {
+    return;
+  }
+  auto durationThreshold = options.durationThreshold.value_or(0.0);
+
+  // observer of type multiple
+  if (options.entryTypes.has_value()) {
+    std::unordered_set<PerformanceEntryType> entryTypes;
+    auto rawTypes = options.entryTypes.value();
+
+    for (auto i = 0; i < rawTypes.size(); ++i) {
+      entryTypes.insert(
+          Bridging<PerformanceEntryType>::fromJs(rt, rawTypes[i]));
+    }
+
+    observer->observe(entryTypes, {.durationThreshold = durationThreshold});
+  } else { // single
+    auto buffered = options.buffered.value_or(false);
+    if (options.type.has_value()) {
+      observer->observe(
+          static_cast<PerformanceEntryType>(options.type.value()),
+          {.buffered = buffered, .durationThreshold = durationThreshold});
+    }
   }
 }
 
-PerformanceEntryReporter::PopPendingEntriesResult
-NativePerformanceObserver::popPendingEntries(jsi::Runtime& /*rt*/) {
-  return PerformanceEntryReporter::getInstance()->popPendingEntries();
-}
+void NativePerformanceObserver::disconnect(
+    jsi::Runtime& rt,
+    jsi::Object observerObj) {
+  auto observer = std::dynamic_pointer_cast<PerformanceObserver>(
+      observerObj.getNativeState(rt));
 
-void NativePerformanceObserver::setOnPerformanceEntryCallback(
-    jsi::Runtime& /*rt*/,
-    std::optional<AsyncCallback<>> callback) {
-  if (callback) {
-    PerformanceEntryReporter::getInstance()->setReportingCallback(
-        [callback = std::move(callback)]() {
-          callback->callWithPriority(SchedulerPriority::IdlePriority);
-        });
-  } else {
-    PerformanceEntryReporter::getInstance()->setReportingCallback(nullptr);
+  if (!observer) {
+    return;
   }
-}
 
-void NativePerformanceObserver::logRawEntry(
-    jsi::Runtime& /*rt*/,
-    const PerformanceEntry entry) {
-  PerformanceEntryReporter::getInstance()->logEntry(entry);
+  observerObj.setNativeState(rt, nullptr);
```

Comment:
I just realized we can't just destroy the observer when we call `disconnect` because you start observing after calling the method. We should create the observer when we instantiate the observer in JS, add it to the registry when observing and removing it from the registry when disconnecting. We should never manually destroy the observer or detach it from the JS object, and just let the JS GC get rid of it when it goes out of scope.